README.md
Ponomar is a fully-functional program suite for the Orthodox Church and provides the following features:

1. Calendar and liturgical information for any day of any year
2. Liturgical readings for the day
3. Lives of saints
4. Liturgical texts in a variety of languages
5. Liturgical service assembly for any day
6. Library of patristic text and scriptural commentary
7. Library of liturgical music in a variety of traditional chant

Copyright 2006-2018 Aleksandr Andreev and others.

Ponomar is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Ponomar is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Ponomar.  If not, see <http://www.gnu.org/licenses/>.

IMPORTANT INFORMATION
Ponomar is ALPHA-PHASE software and is intended for SOFTWARE-TESTING PURPOSES ONLY.

In order to fully use the Ponomar Java program, the following steps can be followed:
1) Install the Ponomar Unicode fonts located at https://sci.ponomar.net/fonts.html. 
Although this font is for primarily displaying Church Slavonic, 
it does contain special Typicon glyphs that are used by most of the other languages.
Note that Ponomar uses the TrueType version of this font (Ponomar Unicode TT)
because Java has poor support of OpenType fonts.

2) Install an SDK for Java. 

3) Create the .class files by typing 

`make`

4) From the root of this project, type 

`java Ponomar.Main`

and the main Ponomar interface should appear.

5) A Perl API is available in Ponomar/APIs/Perl. See its documentation.

6) If you make any changes, be sure to run the regression tests:

`make test`

.

ChangeLog.txt
RELEASE NOTES
Ponomar PRE-RELEASE VERSION 0.0.45 REVISION B
DATE: 03/25/2009 (03/12/2009 old-style)
PLACE: ST PETERSBURG, RUSSIA

This is a minor fix of the Ponomar program, taking care of some outstanding issues.

SUMMARY OF (some of) CHANGES SINCE 0.0.45

File				Line #		Description
------------------------------+---------------+-------------------------------------------------------------------------
Bible.java			616		Bible.main(String) deprecated
Bible.java			497		Fixed a problem whereby reading Psalm 128 made Bible crash
PrintableTextPane.java		35		Rewrote print method fixing a problem whereby if
						more than one page was printed, nothing was printed at all
xml/03/24.xml			4-8		Fixed < and > in xml code to &lt; and &gt;
xml/03/25.xml			4-8		Fixed < and > in xml code to &lt; and &gt;
xml/05/29.xml			15		Fixed error in xml file
xml/Commands/DivineLiturgy.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
xml/Commands/LanguagePacks.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
xml/pentecostarion/65.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
.... to ...
xml/pentecostarion/99.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
xml/pentecostarion/149.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
.... to ...
xml/pentecostarion/315.xml	<all>		Fixed < and > in xml code to &lt; and &gt;
------------------------------------------------------------------------------------------------------------------------


RELEASE NOTES
Ponomar PRE-RELEASE VERSION 0.0.45
DATE: 10/05/2008 (09/22/2008 old-style)
PLACE: ST PETERSBURG, RUSSIA

Another major update of the Ponomar program.
New features (work by Yuri Shardt):
	* Main interface switches between J-ville and Lucan Jump lectionaries
	* Main interface toolbar included
	* Bugs with the lectionary resolved
	* International language support
New features (work by Aleks Andreev):
	* Scripture reader cleaned up, includes parsing of instructions
	* Toolbar included for Scripture reader with Save, Print, etc.

SUMMARY OF (some of) CHANGES SINCE 0.0.40b

File				Line #		Description
------------------------------+---------------+-------------------------------------------------------------------------
Bible.java			32		Toolbar support
				46-47		International language support
				57-59		Toolbar buttons
				74		International language support
				117-149		Toolbar support
				190		Bible() default behaviour changed
				207-226		International language support
				234-237		International language support
				276-412		Toolbar functionality
				563-566		getText(String, String) deprecated
				568-594		Parsing for reading instructions added
				596-611		International language support
				
Helpers.java			108-137		Printing support added

PrintableTextPane.java		<created>	Printing support

Main.java			475		Sunrise-sunset configurability support added
-----------------------------------------------------------------------------------------------------------------------


RELEASE NOTES
Ponomar PRE-RELEASE VERSION 0.0.40 Revision B
DATE: 09/15/2008 (09/02/2008 old-style)
PLACE: ST PETERSBURG, RUSSIA

This minor update fixes a few bugs in the link between the main program interface
and the scripture processing interface.

SUMMARY OF CHANGES SINCE 0.0.40a

File				Line #		Description
------------------------------+---------------+-------------------------------------------------------------------------
Bible.java			62		Bible(string) deprecated

Main.java			411-417		Updated link to Bible, to use Bible(String, String).
------------------------------------------------------------------------------------------------------------------------

RELEASE NOTES
Ponomar PRE-RELEASE VERSION 0.0.40 Revision A
DATE: 08/04/2008 (07/22/2008 old-style)
PLACE: DENVER, COLORADO

This minor update fixes a few bugs in the link between the main program interface
and the Scripture processing interface.
These issues consisted of handling Scripture readings with multiple parts
(e.g. Isa_4:5-10, 16-20) and Scripture readings with spaces in names
(e.g. I Cor_2:3-9).

SUMMARY OF CHANGES SINCE 0.0.40 (initial)

File				Line #		Description
------------------------------+---------------+-------------------------------------------------------------------------
Main.java			433, 446, 459	New call: "<A Href=reading#" + parts[0].replace(' ', '_') + "#" + parts[1].replaceAll(" ", "") + ">";

DivineLiturgy.java		515		Ditto

Bible.java			349, 350	Removed debug code

xml/07/11.xml			9		Fixed typo

TODO: The code at Main.java, lines 433, etc, ought to be rewritten so as not to use
the replaceAll function.
-------------------------------------------------------------------------------------------------------------------------

RELEASE NOTES
Ponomar PRE-RELEASE VERSION 0.0.40 
DATE: 07/02/2008 (06/19/2008 old-style)
PLACE: DENVER, COLORADO

The present update releases an initial Scripture processing interface (Bible.java)
and attempts to implement the Lucan Jump methodology.

SUMMARY OF CHANGES SINCE 0.0.35d

File				Line #		Description
------------------------------+---------------+-------------------------------------------------------------------------
xml/bible.xml			created		This file contains all information
						about Scripture texts used by Ponomar

data/kjv/			created		Implementation of the Orthodox canon
						of Scripture
						(Notice: outside of the UK,
						 this text is in the public domain.
						Its distribution in the UK may be illegal).

**** NOTE: Major changes from the KJV:
	Some books have been renamed to correspond to the Septuagint
	KJV Name	-->	Orthodoxized KJV Name
	I Samuel		I Kingdoms
	II Samuel		II Kingdoms
	I Kings			III Kingdoms
	II Kings		IV Kingdoms
	I Chronicles		I Paralipomenon
	II Chronicles		II Paralipomenon

	Psalms have been renumbered
	KJV Psalm	-->	Orthodoxized KJV Psalm
	1-8			1-8
	9-10			9

Fasting.java
package Ponomar;

import javax.swing.*;
import java.beans.*;
import java.awt.*;
import java.util.*;
import java.io.*;
import javax.swing.event.*;
import java.awt.event.*;
import java.beans.*;
/***********************************************************************
THIS MODULE READS THE FASTING.XML FILE TO DETERMINE THE FAST ON A GIVEN DAY

(C) 2009 YURI SHARDT. ALL RIGHTS RESERVED.

 PERMISSION IS HEREBY GRANTED TO USE, MODIFY, AND/OR REDISTRIBUTE THIS SOURCE CODE
 PROVIDED THAT THIS NOTICE REMAINS IN ALL VERSION AND / OR DERIVATIVES THEREOF.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
 OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
***********************************************************************/

public class Fasting implements DocHandler
{
	private final static String FileName   = "Ponomar/xml/Commands/Fasting.xml";
	private static boolean readPeriod=false;
	private static boolean readLanguage=false;
	private static OrderedHashtable Information;
	
	private LanguagePack Phrases=new LanguagePack();
	//GET THE APPROPRIATE FASTING LINES
	//private String[] ServiceNames=Text.obtainValues((String)Text.Phrases.get("ServiceRead"));
	//private String[] LanguageNames=Text.obtainValues((String)Text.Phrases.get("LanguageMenu"));
	private static String Fast;
	
	public String FastRules() //throws IOException
	{
		Fast=new String();
		Information=new OrderedHashtable();
		//THIS IS A KLUTZ THAT WILL BE REMOVED ONCE THERE IS A PROPER ABILITY TO RANK THE DAY
		//RANK 1 HOLIDAYS
		StringOp.dayInfo.put("dRank",5);	//ANY RANK LOWER THAN 4 WILL DO
		try
		{
			FileReader frf = new FileReader("Ponomar/xml/Commands/DivineLiturgy.xml");
			QDParser.parse(this, frf);
			int doy=Integer.parseInt(StringOp.dayInfo.get("doy").toString());
			int nday=Integer.parseInt(StringOp.dayInfo.get("nday").toString());
			if(doy == 256 || doy == 358 || doy == 5 ||  nday == 40 || nday == 50 || doy == 217)
			{
				//THESE ARE CLASS 1 HOLIDAYS
				StringOp.dayInfo.put("dRank",1);
			}
			if ((doy == 32 && nday != -48) || doy == 250 || doy == 226 || doy == 324 || doy == 83)
			{
				//THESE ARE CLASS 2 HOLIDAYS
				StringOp.dayInfo.put("dRank",2);
			}
			//THIS WILL NOT CATCH ALL THE RANK  3 OR 4 HOLIDAYS, BUT MOST OF THEM WILL BE SO CAUGHT
			Vector vect = (Vector) Information.get("Class3Transfers");
			if(vect != null)
			{
				for(Enumeration e2=vect.elements();e2.hasMoreElements();)
				{	
					String Command = (String)e2.nextElement();
					if (StringOp.evalbool(Command))
					{
						StringOp.dayInfo.put("dRank",3);	//IT DOES NOT MATTER FOR THE FASTING RULES WHETHER IT IS 3 OR 4 AS THE RANK.
					}
				}
			}
			//ADDING SOME FURTHER DAYS THAT ARE NOT INCLUDED ABOVE
			if(doy == 6)
			{
				StringOp.dayInfo.put("dRank",3);
			}
			if(doy == 127 || doy == 128 || doy == 161 || doy == 169 || doy == 185 || doy == 190 || doy == 199 || doy == 200 || doy == 220 || doy == 227 || doy == 239 || doy == 241 || doy == 242 || doy == 267 || doy == 281 || doy == 282 || doy == 290 || doy == 291 || doy == 295 || doy == 300 || doy == 328 || doy == 342 || doy == 345 || doy == 346 || doy == 350 || doy == 23 || doy == 26 || doy == 16 || doy == 19 || doy == 8 || doy == 10 || doy == 11 || doy == 13 || doy == 24)
			{
				StringOp.dayInfo.put("dRank",4);
			}
						
		}
		catch (Exception e1)
		{
			e1.printStackTrace();
		}
		//FINISHED THE KLUTZ
		//ACTUAL PROGRAMME
		//System.out.print("Today's rank is "+StringOp.dayInfo.get("dRank")+"\n");
		try
		{
			BufferedReader frf1 = new BufferedReader(new InputStreamReader(new FileInputStream(FileName), "UTF8"));
			QDParser.parse(this, frf1);
		}
		catch (Exception e)
		{
			e.printStackTrace();
			return null;			//THERE WAS AN ERROR IN PROCESSING THE FILES
		}
		//NOW IT IS NECESSARY TO CONVERT THE COMPUTER SPEAK TO HUMAN SPEAK
		return convert(Fast);
	}
	public String convert(String Fast)
	{
		/*COMMON REGULATIONS
		0000000 : No Food
		0000001 : Strict Fast (Dry Food Only)
		0000011 : Food without oil
		0000111 : Food with oil/Fast: Wine and oil allowed
		0001111 : Caviar is permitted 
		0011111 : Fish is permitted/Fast: Fish, wine, and oil allowed
		0111111 : Meat is excluded
		1111111 : No fast
		0000010 : Wine Permitted
		*/
		
		String[] FastNames= Phrases.obtainValues((String)Phrases.Phrases.get("Fasts"));
				
		if(Fast.equals("0000000"))
		{
			return FastNames[4]+" " + FastNames[1]+FastNames[6]+" "+ FastNames[7];
		}
		if(Fast.equals("0000001"))
		{
			return  FastNames[4]+" " +FastNames[1]+FastNames[6]+" "+ FastNames[8];
		}
		if(Fast.equals("0000011"))
		{
			return  FastNames[4]+" " +FastNames[1]+FastNames[6]+" "+ FastNames[9];
		}
		if(Fast.equals("0000111"))
		{
			return  FastNames[4]+" " +FastNames[2];
		}
		if(Fast.equals("0001111"))
		{
			return FastNames[4]+" " +FastNames[10];
		}
		if(Fast.equals("0011111"))
		{
			return  FastNames[4]+" " +FastNames[3];
		}
		if(Fast.equals("0111111"))
		{
			return FastNames[4]+" " +FastNames[11];
		}
		if(Fast.equals("1111111"))
		{
			return  FastNames[4]+" " +FastNames[0];
		}
		if(Fast.equals("0000010"))
		{
			return FastNames[4]+" " +FastNames[12];
		}
		//NONE OF THE PREDEFINED SEQUENCES WERE ENCOUNTERED.
		//PARSE IT ELEMENT BY ELEMENT!
		String[] item = new String[] {FastNames[13],FastNames[14],FastNames[15],FastNames[16],FastNames[17],FastNames[18],FastNames[19]};
		String[] permitted=new String[7];
		String[] forbidden=new String[7];
		int permit=0;
		int forbid=0;
		for(int i=0;i < 7;i++)
		{
			if(Fast.substring(i,i+1).equals("1"))
			{
				//THE GIVEN ITEM IS PERMITTED
				permitted[permit]=item[i];
				permit++;
			}
			else
			{
				//THE GIVEN ITEM IS FORBIDDEN
				forbidden[forbid]=item[i];
				forbid++;
			}
		}
		String output=FastNames[27]+" ";
		for(int i=0;i < permit;i++)
		{
			output+=permitted[i];
			if(i<permit-1)
			{
				output+=", ";
			}
			if(i==permit-2)
			{
				output+=" "+FastNames[25]+" ";
			}
		}
		if(permit==1)
		{
			output+=" "+FastNames[20]+" "+FastNames[23]; 
		}
		else if(permit==2)
		{
			//FOR THOSES SLAVIC LANGUAGES WITH THE DUAL
			output+=" "+FastNames[21]+" "+FastNames[23];
		}
		else
		{
			output+=" "+FastNames[22]+" "+FastNames[23];
		}
		output+=FastNames[26]+" ";
		for(int i=0;i < forbid;i++)
		{
			output+=forbidden[i];
			if(i<forbid-1)
			{
				output+=", ";
			}
			if(i==forbid-2)
			{
				output+=" "+FastNames[25]+" ";
			}
		}
		if(forbid==1)
		{
			output+=output+=" "+FastNames[20]+" "+FastNames[24]; 
		}
		else if(forbid==2)
		{
			//FOR THOSES SLAVIC LANGUAGES WITH THE DUAL
			output+=" "+FastNames[21]+" "+FastNames[24];
		}
		else
		{
			output+=" "+FastNames[22]+" "+FastNames[24];
		}
		return FastNames[4]+" " +output;
	}
				
	public void startDocument()
	{

	}

	public void endDocument()
	{

	}

	public void startElement(String elem, Hashtable table)
	{
		
		// THE TAG COULD CONTAIN A COMMAND Cmd
		// THE COMMAND TELLS US WHETHER OR NOT TO PROCESS THIS TAG GIVEN
		// TODAY'S INFORMATION IN dayInfo. 
		if (table.get("Cmd") != null)
		{
			// EXECUTE THE COMMAND, AND STOP IF IT IS FALSE
			
			if (StringOp.evalbool(table.get("Cmd").toString()) == false) 
			{
				return;
			}
		}
		if(elem.equals("LANGUAGE"))
		{
			readLanguage=true;
		}
		if(elem.equals("PERIOD"))
		{
			readPeriod=true;
		}
		if(elem.equals("RULE") && readPeriod && readLanguage)
		{
			//A POTENTIAL FASTING RULE HAS BEEN ENCOUNTERED THAT APPLIES FOR TODAY.
			Fast=table.get("Case").toString();
		}
		//THE FOLLOWING SECTION SHOULD BE REMOVED ONCE THERE IS A PROPER RANKING OF DAYS
		if (elem.equals("COMMAND"))
		{
		//THIS WILL STORE ALL THE POSSIBLE COMMANDS FOR A GIVEN SITUATION AND ALLOW THE RESULTS TO BE DETEMINED.
		String name = table.get("Name").toString();
		String value=table.get("Value").toString();
		//IF THE GIVEN name OCCURS IN THE information HASHTABLE THAN AUGMENT ITS VALUES.
		if (Information.containsKey(name))
		{
			Vector previous = (Vector)Information.get(name);
			previous.add(value);
			Information.put(name,previous);
		}
		else
		{
			Vector vect = new Vector();
			vect.add(value);
			Information.put(name,vect);
		}
		
		}
		//TO HERE REMOVE

	}

	public void endElement(String elem)
	{
		if(elem.equals("LANGUAGE"))
		{
			readLanguage=false;
		}
		if(elem.equals("PERIOD"))
		{
			readPeriod=false;
		}
	}

	public void text(String text)
	{

	}

	private boolean eval(String expression) throws IllegalArgumentException
	{
		return false;
	}
	
}